### PR TITLE
Add sys_flock and sys_fcntl

### PR DESCRIPTION
**Pull request by grief8**
## Summary

This PR introduces the implementation of `sys_flock` and `sys_fcntl` system calls to enhance file locking capabilities in Asterinas. The addition of these calls enables processes to perform advisory file locking, allowing for better coordination and synchronization when accessing shared files among multiple processes.

## Background

Advisory locking is a technique that allows processes to lock sections of a file to prevent concurrent access, which could otherwise lead to data corruption or inconsistency. 

### sys_flock

- Added `sys_flock()` function to handle the `F_LOCK` and `F_UNLOCK` operations based on the POSIX standard.
- Implemented lock management to maintain a list of locked regions of files and ensure that they don't overlap for the same file descriptor by different processes.

### sys_fcntl

- Enhanced `sys_fcntl()` to support `F_SETLK`, `F_SETLKW`, and `F_GETLK` commands for file locking.
- Added helper functions to manage the complexity of converting between user-space and kernel-space data structures for lock descriptions.
- Provided support for handling non-blocking (asynchronous) lock attempts by allowing the system call to return if the lock acquisition is not possible immediately.

## Testing

- Added unit tests for both `sys_flock()` and `sys_fcntl()` to validate their expected behavior in various scenarios, including locking, unlocking, and concurrent lock contention.
- Performed integration testing to ensure that the locking mechanisms work seamlessly with existing file system operations and multiple processes.

## Impact

With the incorporation of these new system calls, our operating system now supports robust file locking mechanisms, which will be beneficial for applications that rely on file-based coordination and synchronization. This also brings our system closer to POSIX compliance with respect to file locking.

### CODE REVIEW

**Comment by grief8** [comment]
Hi @liqinggd , could you please take a moment to review my PR? I've made the implementation on your early design and would appreciate your feedback. Thanks in advance!

**Comment by liqinggd** [comment]
Can we divide the commit into two commits? one is for `sys_flock` and the other is for `fcntl`.

**Comment by grief8** [comment]
Hi @liqinggd, thank you for your feedback! I have split the PR into two commits. Hoping for your comments at your convenience.

File: kernel/aster-nix/src/fs/utils/range_lock/range.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+#[derive(Debug, Copy, Clone)]
+#[repr(u16)]
+pub enum RangeLockWhence {
+    SEEK_SET = 0,
+    SEEK_CUR = 1,
+    SEEK_END = 2,
+}
+
+impl RangeLockWhence {
+    pub fn from_u16(whence: u16) -> Result<Self> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/utils/range_lock/range.rs:176]: You can derive `TryFromInt` trait to convert a u16 to `RangeLockWhence`

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 2]
```diff
@@ -61,6 +64,67 @@ impl InodeHandle<Rights> {
         }
         self.0.readdir(visitor)
     }
+
+    pub fn position(&self) -> Result<i64> {
+        Ok(self.0.offset() as i64)
+    }
+
+    pub fn test_advisory_lock(&self, lock: &mut RangeLock) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:72]: I prefer to use `test_range_lock` here

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 3]
```diff
@@ -61,6 +64,67 @@ impl InodeHandle<Rights> {
...
+                lock.set_type(RangeLockType::F_UNLCK);
+            }
+            Some(range_lock_list) => {
+                range_lock_list.test_lock(lock);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn set_advisory_lock(&self, lock: &RangeLock, is_nonblocking: bool) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:94]: I prefer to use `set_range_lock` here

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 4]
```diff
@@ -61,6 +64,67 @@ impl InodeHandle<Rights> {
...
+        };
+        let range_lock_list = match extension.get::<RangeLockList>() {
+            Some(list) => list,
+            None => extension.get_or_put_default::<RangeLockList>(),
+        };
+
+        range_lock_list.set_lock(lock, is_nonblocking)
+    }
+
+    pub fn release_advisory_locks(&self) {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:119]: I prefer to use `release_range_locks`

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 5]
```diff
@@ -191,6 +192,40 @@ impl InodeHandle_ {
 
         self.dentry.inode().ioctl(cmd, arg)
     }
+
+    fn check_advisory_lock_with_access_mode(&self, lock: &RangeLock) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:196]: `check_range_lock_with_access_mode`

**Comment by liqinggd** [comment]
Here is the manual of range lock:
```
If a process closes any file descriptor referring to a file,
          then all of the process's locks on that file are released,
          regardless of the file descriptor(s) on which the locks were
          obtained. 
```

So we should release the locks when file is closed.

File: kernel/aster-nix/src/fs/utils/inode.rs [hunk 6]
```diff
@@ -361,6 +361,11 @@ pub trait Inode: Any + Sync + Send {
     fn is_dentry_cacheable(&self) -> bool {
         true
     }
+
+    /// Get the extension of this inode
+    fn extension(&self) -> Option<&Extension> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/utils/inode.rs:366]: Use `fn extension(&self) -> &Extension` istead

File: kernel/aster-nix/src/fs/ext2/inode.rs [hunk 7]
```diff
@@ -28,6 +29,7 @@ pub struct Inode {
     block_group_idx: usize,
     inner: RwMutex<Inner>,
     fs: Weak<Ext2>,
+    extension: Option<Extension>,
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/ext2/inode.rs:32]: You are right!

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 8]
```diff
@@ -61,6 +64,67 @@ impl InodeHandle<Rights> {
...
+        let range_lock = RangeLockBuilder::new()
+            .type_(RangeLockType::F_UNLCK)
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:121]: I think `type_` may provide better readability🤔

File: kernel/aster-nix/src/fs/utils/range_lock/mod.rs [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+#[derive(Debug, Copy, Clone, PartialEq)]
+#[repr(u16)]
+pub enum RangeLockType {
+    F_RDLCK = 0,
+    F_WRLCK = 1,
+    F_UNLCK = 2,
+}
+
+impl RangeLockType {
+    pub fn from_u16(_type: u16) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/range_lock/mod.rs:375]: That's really a good idea!

File: kernel/aster-nix/src/fs/utils/inode.rs (hunk 6 above)
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/inode.rs:366]: Changed.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 3 above)
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:94]: Thank you for your suggestions. I will rename them.

**Comment by liqinggd** [comment]
Let's merge the two commits first, so it will be easier for you to make modifications. After we complete the code review, you can split the commit into two again.

File: kernel/aster-nix/src/fs/utils/inode.rs [hunk 10]
```diff
@@ -375,6 +375,11 @@ pub trait Inode: Any + Sync + Send {
...
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/utils/inode.rs:380]: Hi， I think we should use `fn extension(&self) -> &Extension` here.

**Comment by grief8** [comment]
> Let's merge the two commits first, so it will be easier for you to make modifications. After we complete the code review, you can split the commit into two again.

All right! Thank you for your patience!

File: kernel/aster-nix/src/fs/file_table.rs [hunk 11]
```diff
@@ -174,6 +177,9 @@ impl FileTable {
     }
 
     pub fn get_entry(&self, fd: FileDesc) -> Result<&FileTableEntry> {
+        if fd < 0 {
+            return_errno_with_message!(Errno::EBADF, "Invalid file descriptor: fd is negative");
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/file_table.rs:181]: To maintain code consistency, do not capitalize the first letter here.

File: kernel/aster-nix/src/fs/file_table.rs [hunk 12]
```diff
@@ -160,6 +160,9 @@ impl FileTable {
     }
 
     pub fn get_file(&self, fd: FileDesc) -> Result<&Arc<dyn FileLike>> {
+        if fd < 0 {
+            return_errno_with_message!(Errno::EBADF, "Invalid file descriptor: fd is negative");
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/file_table.rs:164]: To maintain code consistency, do not capitalize the first letter here.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 13]
```diff
@@ -61,6 +64,101 @@ impl InodeHandle<Rights> {
         }
         self.0.readdir(visitor)
     }
+
+    pub fn position(&self) -> Result<i64> {
+        Ok(self.0.offset() as i64)
+    }
+
+    pub fn test_range_lock(&self, lock: &mut RangeLock) -> Result<()> {
+        let extension = match self.dentry().inode().extension() {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:73]: If the `extension` method does not return an `Option`, the code can be simplified.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (liqinggd) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:121]: Fine

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 14]
```diff
@@ -126,6 +224,8 @@ impl FileLike for InodeHandle<Rights> {
 
     fn clean_for_close(&self) -> Result<()> {
         // Close does not guarantee that the data has been successfully saved to disk.
+        self.release_range_locks();
+        self.unlock_flock();
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:228]: I think the `unlock_flock()` is called when dropping the `InodeHandle`. 
The `clean_for_close` is called when we close a file, but the `InodeHandle` may not be dropped. This is a **big** difference.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 15]
```diff
@@ -191,6 +192,40 @@ impl InodeHandle_ {
 
         self.dentry.inode().ioctl(cmd, arg)
     }
+
+    fn check_range_lock_with_access_mode(&self, lock: &RangeLock) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:196]: This function should be implemented in `dyn_cap.rs` to do the permission check.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 16]
```diff
@@ -61,6 +64,101 @@ impl InodeHandle<Rights> {
         }
         self.0.readdir(visitor)
     }
+
+    pub fn position(&self) -> Result<i64> {
+        Ok(self.0.offset() as i64)
+    }
+
+    pub fn test_range_lock(&self, lock: &mut RangeLock) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:72]: I think we should move these new code into `mod.rs`, and the code in `dyn_cap.rs` just do some permission check and call the functions in `mode.rs`, this is the design of `InodeHandle`.

File: kernel/aster-nix/src/fs/utils/inode.rs (hunk 10 above)
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/inode.rs:380]: Although most filesystem have locks, I find that some filesystem may not support `fcntl` and `flock` like `sysfs` and `procfs`. Therefore, `Option` still needs to be remained.

A simple test can be carried on:
```sh
#!/bin/bash

# Select a file in sysfs to test
TEST_FILE="/sys/kernel/debug/tracing/trace_marker"

# Check if the file exists
if [ ! -e "$TEST_FILE" ]; then
  echo "Test file does not exist: $TEST_FILE"
  exit 1
fi

# Try using flock to lock the file
(
  sudo flock -x 200 || exit 1
  echo "File locked successfully"
  sleep 10
) 200>"$TEST_FILE"

# Check if the lock is successful
if [ $? -eq 0 ]; then
  echo "Locking succeeded"
else
  echo "Locking failed"
fi
```

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 14 above)
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:228]: That's more reasonable. 

For `fcntl`:
> If a process closes any file descriptor referring to a file, then all of the process's locks on that file are released, regardless of the file descriptor(s) on which the locks were obtained.

For `flock`:
> lock is released either by an explicit LOCK_UN operation on any of these duplicate file descriptors, or when all such file descriptors have been closed.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 17]
```diff
@@ -191,6 +192,106 @@ impl InodeHandle_ {
...
+        if let Some(extension) = self.dentry.inode().extension() {
+            let flock_list = match extension.get::<FlockList>() {
+                Some(list) => list,
+                None => extension.get_or_put_default::<FlockList>(),
+            };
+
+            flock_list.set_lock(lock, is_nonblocking)
+        } else {
+            warn!("Inode extension is not supported, let the lock could be acquired");
+            // TODO: Implement inode extension for FS
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:256]: Please remove the `TODO`.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 18]
```diff
@@ -191,6 +192,106 @@ impl InodeHandle_ {
...
+        if let Some(extension) = self.dentry.inode().extension() {
+            let range_lock_list = match extension.get::<RangeLockList>() {
+                Some(list) => list,
+                None => extension.get_or_put_default::<RangeLockList>(),
+            };
+
+            range_lock_list.set_lock(lock, is_nonblocking)
+        } else {
+            warn!("Inode extension is not supported, let the lock could be acquired");
+            // TODO: Implement inode extension for FS
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:229]: Please remove the `TODO`.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 19]
```diff
@@ -191,6 +192,106 @@ impl InodeHandle_ {
...
+        if let Some(extension) = self.dentry.inode().extension() {
+            if let Some(range_lock_list) = extension.get::<RangeLockList>() {
+                range_lock_list.test_lock(lock);
+            } else {
+                // The range lock could be placed if there is no lock list
+                lock.set_type(RangeLockType::F_UNLCK);
+            }
+        } else {
+            warn!("Inode extension is not supported, the lock could be placed");
+            // Some file systems may not support flock like procfs and sysfs
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:206]: It is "range lock", not "flock".

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 20]
```diff
@@ -191,6 +192,106 @@ impl InodeHandle_ {
...
+    fn test_range_lock(&self, lock: &mut RangeLock) -> Result<()> {
+        if let Some(extension) = self.dentry.inode().extension() {
+            if let Some(range_lock_list) = extension.get::<RangeLockList>() {
+                range_lock_list.test_lock(lock);
+            } else {
+                // The range lock could be placed if there is no lock list
+                lock.set_type(RangeLockType::F_UNLCK);
+            }
+        } else {
+            warn!("Inode extension is not supported, the lock could be placed");
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/inode_handle/mod.rs:205]: I think the `warn` canbe replaced by `debug`.

**Comment by liqinggd** [comment]
LGTM

File: kernel/aster-nix/src/fs/utils/flock.rs [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+};
+
+/// Kernel representation of FLOCK
+pub struct Flock {
+    /// Owner of FLOCK, an opened file descriptor holding the lock
+    owner: Weak<dyn FileLike>,
+    /// Type of lock, SH_LOCK or EX_LOCK
+    type_: FlockType,
+    /// Optional waiters that are blocking by the lock
+    waitqueue: Option<WaitQueue>,
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/utils/flock.rs:20]: See the other comments about `RangeLock` as to why having a `Option<WaitQueue>` field is a bad idea.

File: kernel/aster-nix/src/fs/utils/flock.rs [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+        } else if ops.contains(FlockOps::LOCK_SH) {
+            Self::SH_LOCK
+        } else {
+            panic!("invalid flockops");
+        }
+    }
+}
+
+bitflags! {
+    pub struct FlockOps: i32 {
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/utils/flock.rs:173]: Again, move C definitions to the `syscall` module.

File: kernel/aster-nix/src/fs/utils/flock.rs [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    pub fn set_lock(&self, mut req_lock: Flock, is_nonblocking: bool) -> Result<()> {
+        debug!(
+            "set_lock with Flock: {:?}, is_nonblocking: {}",
+            req_lock, is_nonblocking
+        );
+
+        loop {
+            let mut list = self.inner.write();
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/utils/flock.rs:100]: The same deadlock problem as `RangeLockList`.

**Comment by grief8** [comment]
@boterinas author

File: kernel/aster-nix/src/fs/procfs/template/dir.rs [hunk 24]
```diff
@@ -186,6 +186,10 @@ impl<D: DirOps + 'static> Inode for ProcDir<D> {
     fn is_dentry_cacheable(&self) -> bool {
         !self.common.is_volatile()
     }
+
+    fn extension(&self) -> Option<&Extension> {
+        Some(&self.common.extension)
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/procfs/template/dir.rs]: You are right. We can just leave it unimplemented and the lock will return a success as Linux.

File: kernel/aster-nix/src/fs/file_table.rs [hunk 25]
```diff
@@ -166,6 +166,9 @@ impl FileTable {
     }
 
     pub fn get_file(&self, fd: FileDesc) -> Result<&Arc<dyn FileLike>> {
+        if fd < 0 {
+            return_errno_with_message!(Errno::EBADF, "invalid file descriptor: fd is negative");
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/file_table.rs]: Yes, we can do it in another PR and proposes RawFd to handle negative `FileDesc` like `AT_FDCWD`.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs [hunk 26]
```diff
@@ -191,6 +192,80 @@ impl InodeHandle_ {
...
+            let range_lock_list = match extension.get::<RangeLockList>() {
+                Some(list) => list,
+                None => extension.get_or_put_default::<RangeLockList>(),
+            };
+
+            range_lock_list.set_lock(lock, is_nonblocking)
+        } else {
+            debug!("Inode extension is not supported, let the lock could be acquired");
+            // Some file systems may not support range lock like procfs and sysfs
+            // Returns Ok if extension is not supported.
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/mod.rs:231]: The latter. Ignore the fact that file locking is unsupported and return a success.

File: kernel/aster-nix/src/fs/utils/range_lock/mod.rs [hunk 27]
```diff
@@ -0,0 +1,372 @@
...
+
+mod builder;
+mod range;
+
+/// File POSIX advisory range locks
+
+/// C struct for a file range lock in Libc
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Pod)]
+pub struct c_flock {
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/range_lock/mod.rs]: Thank you for your suggestions! I will modify them in the newest version.

File: kernel/aster-nix/src/fs/utils/range_lock/mod.rs [hunk 28]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+impl RangeLockList {
+    pub fn new() -> Self {
+        Self {
+            inner: RwLock::new(VecDeque::new()),
+        }
+    }
+
+    pub fn test_lock(&self, lock: &mut RangeLock) {
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/range_lock/mod.rs]: Get it! Now it never needs a `&mut`.

File: kernel/aster-nix/src/fs/utils/range_lock/mod.rs [hunk 29]
```diff
@@ -0,0 +1,372 @@
...
+        lock.type_ = RangeLockType::F_UNLCK;
+    }
+
+    pub fn set_lock(&self, lock: &RangeLock, is_nonblocking: bool) -> Result<()> {
+        debug!(
+            "set_lock with RangeLock: {:?}, is_nonblocking: {}",
+            lock, is_nonblocking
+        );
+        loop {
+            let mut list = self.inner.write();
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/range_lock/mod.rs]: Thanks for your reminder! I will replace this implementation with thread-safe ones.

File: kernel/aster-nix/src/fs/utils/range_lock/mod.rs [hunk 30]
```diff
@@ -0,0 +1,372 @@
...
+/// Kernel representation of file range lock
+pub struct RangeLock {
+    /// Owner of lock, process holding the lock
+    owner: Pid,
+    /// Type of lock, F_RDLCK, F_WRLCK, or F_UNLCK
+    type_: RangeLockType,
+    /// Range of lock
+    range: FileRange,
+    /// Waiters that are blocking by the lock
+    waitqueue: Option<WaitQueue>,
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/utils/range_lock/mod.rs]: Great. It seems better.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs [hunk 31]
```diff
@@ -71,6 +71,10 @@ impl InodeHandle<Rights> {
         }
         self.0.readdir(visitor)
     }
+
+    pub fn position(&self) -> Result<i64> {
+        Ok(self.0.offset() as i64)
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:76]: `offset` is usize, and `position` can be a negative value. I think it should be distinguished, so I changed the name.

**Comment by grief8** [comment]
@boterinas ready

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:76]: Oh. I even didn't notice the return value is signed. Why are negative positions even allowed?

File: kernel/aster-nix/src/fs/inode_handle/mod.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/aster-nix/src/fs/inode_handle/mod.rs:231]: If it returns success, how do you know file locking is unsupported?

File: kernel/aster-nix/src/fs/file_table.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/aster-nix/src/fs/file_table.rs]: Could you add an issue?

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:76]: Indeed, `position()` returns `i64` primarily to harmonize with `c_flock`, as the POSIX standard conventionally employs `i64` to signify the file offset. And `fcntl` should process negative values. This is explicitly outlined in the [documentation](https://man7.org/linux/man-pages/man2/fcntl.2.html) for `fcntl`.
>l_start can be a negative number provided the offset does not lie before the start of the file.
>POSIX.1-2001 allows (but does not require) an implementation to support a negative l_len value

File: kernel/aster-nix/src/fs/inode_handle/mod.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/mod.rs:255]: `fcntl` has a command `F_GETLK` to check the status of the lock. If the lock fails, `F_GETLK` will find that the lock is still unlocked. `flock` can be confirmed by a new process trying to acquire the lock.

File: kernel/aster-nix/src/fs/file_table.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/file_table.rs]: Of course. I have added #1146 for this now.

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs:76]: I have explained this point to you in this PR, just at a different location: the `fs` module should be Rust-style kernel types, whereas the `syscall` module should implement Linux-compatible C types. We don't want to sacrifice type safety for Linux compatibility reasons in `fs` module. In this particular case, it is impossible for the file cursor of a `InodeFile` to be negative, right? Then, we should not use `i64`. Take care of the type conversion (from `usize` to `i64`) in the `syscall` module or the `fcntl` implementation.

> l_start can be a negative number provided the offset does not lie before the start of the file.

Yes, that is what the manual says. But the manual also says:

>       l_start is the starting offset for the lock, and is interpreted relative to either: the start of the file (if l_whence is SEEK_SET); the current file offset (if l_whence is SEEK_CUR); or the end of the file (if l_whence is SEEK_END).  In the final two cases, l_start can be a negative number provided the offset does not lie before the start of the file.

Basically, `l_start` can be a relative value depending on the value of `l_whence`. So of course, it should can be negative and use `i64`.

File: kernel/aster-nix/src/fs/inode_handle/mod.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/aster-nix/src/fs/inode_handle/mod.rs:255]: Sounds reasonable.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

File: kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/aster-nix/src/fs/inode_handle/dyn_cap.rs]: Get it! Thank you for your instructions! I have moved the type conversion to `syscall` module.

**Comment by grief8** [comment]
@boterinas rerun
